developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,349 @@
+contain a new token, referencing connection point components in the host
+topology. This value is then assigned to the .ref_comp_id field of struct
+sof_ipc_comp_config during topology file parsing and sent by the guest to the
+host. The host then uses that value to modify SOF_IPC_TPLG_COMP_CONNECT
+connection IPC messages, involving connection points, from the guest before
+sending them to the DSP.
+
+3.2. Streaming
+--------------
+
+Guest audio streaming is mostly transparent for the host. The host audio
+subsystem doesn't get involved with most guest streaming or kcontrol operations
+because the virtualisation is applied at the DSP level, not at the audio level.
+All those operations are completely processed on the guest and only resulting
+DSP IPCs are forwarded to the host. The only exceptions are beginning and ending
+of the streaming, at which times the ASoC subsystem on the host has to be
+informed, that a certain pipeline is becoming active or inactive respectively.
+
+The SOF vhost driver uses the STREAM_PCM_PARAMS IPC message from the guest to
+allocate an ALSA PCM runtime object and configure the HDA hardware. If zero-copy
+is not used and audio data is transferred via VirtQueues, DMA buffers and also
+allocated.
+
+Upon reception of the STREAM_TRIG_START IPC message the vhost driver activates
+the associated pipeline and updates the DPCM routing information on the host.
+
+As mentioned above, it is our goal to implement data zero-copy. As long as
+this isn't the case, the guest VirtIO driver implements .block_write() and
+.block_read() SOF DSP operation methods for data streaming via the data
+VirtQueue.
+
+The host receives requests on the data VirtQueue and copies data between PCM
+runtime buffers and the queue and responds either with a status or with a data
+buffer.
+
+Similarly to the data, streaming buffer position updates are currently also
+transferred via a dedicated VirtQueue. Both the host and the guest are
+configured to use IPC messages for position updates. Buffers for VirtIO
+VirtQueues are always provided by guests, therefore there should always be a
+position update buffer waiting on the host side to send a position update
+message as soon as one arrives from the DSP.
+
+Upon reception of the STREAM_TRIG_STOP IPC message the vhost driver updates
+the DPCM routing information and deactivates the virtual PCM pipeline, described
+in the `2.2. Topology`_ section above, on the host. This operation requires
+particular care, because in the present state, just calling
+
+|	soc_dpcm_runtime_update();
+|	snd_soc_runtime_deactivate();
+
+doesn't deactivate the pipeline and therefore doesn't allow runtime PM to
+suspend the interface, which then breaks following activation attempts. The
+current SOF VirtIO implementation contains a fix for that, which has to be
+upstreamed along with the rest of the ALSA core virtualisation modifications
+and extensions.
```

Comment:
runtime_pm on the host only works with open() and close() operations. It doesn't matter if there is a trig_start/stop, the refcount is only based on open/close. I see danger here if we use different concepts for virtualization. it should be when the PCM stream is closed that pm_runtime can kick in (and with a N s delay).

Reply:
@plbossart this isn't about runtime PM, it's about audio runtimes / pipelines. It affects the runtime PM but only indirectly. As I mentioned in my today's comment to https://github.com/thesofproject/linux/pull/1501 earlier today, I now have a version of runtime PM, which works synchronously to the guest rt-pm. A description of runtime-PM in this document is mostly described in section 3.3 below. If we agree to use the new SOF VirtIO runtime-PM concept, I'll update that chapter too.